tests/xmap_test.py

```diff
@@ -335,6 +532,47 @@ def f(x, y):
 
     self.assertAllClose(z, jnp.einsum('nij,njk->nik', x, y))
 
+  @parameterized.named_parameters(jtu.cases_from_list(
+      {"testcase_name": f"_{next(test_counter)}",
+       "lhs_shape": lhs_shape, "rhs_shape": rhs_shape, "spec": spec,
+       "axis_resources": axis_resources, "mesh_data": mesh_data}
+      for test_counter in [it.count()]
+      for lhs_shape, rhs_shape in it.product(
+          [(2, 2), (2, 3, 2), (2, 3, 1, 2)],
+          repeat=2)
+      for spec in all_pdot_specs(lhs_shape, rhs_shape)
+      for axis_resources, mesh_data in schedules_from_pdot_spec(
+          spec, lhs_shape, rhs_shape)))
+  @ignore_xmap_warning()
+  def testPdotSystematic(self, lhs_shape, rhs_shape, spec, axis_resources,
+                         mesh_data):
+    names = gen_axis_names(spec.num_mapped)
+    contract_names, batch_names = split_list(names, [spec.num_mapped_contract])
+    lhs_spec = dict(zip(spec.lhs_mapped, names))
+    rhs_spec = dict(zip(spec.rhs_mapped, names))
```

Comment:
This part is a little confusing, because it depends on some unsaid invariants of `PdotSpec` (in this case that `lhs_mapped` returns `(*map_contract, *map_batch)`. It would be nicer if we could put this spec generation in the `PdotSpec` itself, so that one doesn't have to remember about those properties to understand this bit of code.

Reply:
Hmm, well right now `PdotSpec` doesn't include any name information. But we could factor the names into it, which might help simplify this calling code as well as other things. I'll give that a shot.